app/assets/javascripts/labbook.coffee

```diff
@@ -74,6 +79,12 @@ class LabbookController
   showDialog: ->
     @$dialog.dialog('open')
 
+  onDialogClosing: (event) ->
+    if @implementsIFramePhone and @hasUnsavedState
```

Comment:
Instead of relying on the labbook to keep its state up-to-date, can you just have LARA ask the iframe for its state here, and block until it gets a response if the labbook implements iframe phone.  Then it will be the labbooks responsibility for telling the user about unsaved state . This will mean you'll need a way to handle dialogClosing I'm not sure if the dialog library supports that.